tower/src/buffer/future.rs

```diff
@@ -4,47 +4,87 @@ use super::{error::Closed, message};
 use futures_core::ready;
 use pin_project::pin_project;
 use std::{
+    fmt::Debug,
     future::Future,
     pin::Pin,
     task::{Context, Poll},
 };
 
 /// Future that completes when the buffered service eventually services the submitted request.
 #[pin_project]
-#[derive(Debug)]
-pub struct ResponseFuture<T> {
+pub struct ResponseFuture<S, E2, Response>
+where
+    S: crate::Service<Response>,
+{
     #[pin]
-    state: ResponseState<T>,
+    state: ResponseState<S, E2, Response>,
+}
+
+impl<S, E2, Response> Debug for ResponseFuture<S, E2, Response>
+where
+    S: crate::Service<Response>,
+    S::Future: Debug,
+    S::Error: Debug,
+    E2: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ResponseFuture")
+            .field("state", &self.state)
+            .finish()
+    }
 }
 
 #[pin_project(project = ResponseStateProj)]
-#[derive(Debug)]
-enum ResponseState<T> {
-    Failed(Option<crate::BoxError>),
-    Rx(#[pin] message::Rx<T>),
-    Poll(#[pin] T),
+enum ResponseState<S, E2, Response>
+where
+    S: crate::Service<Response>,
+{
```

Comment:
Take it or leave it: I think that if this struct was generic over type parameters for the service's future and error types, rather than being generic over a `Service`, it would still be able to derive `Debug` (it's the `Response` type we don't care about implementing `Debug`), while still having the same number of type parameters (`T, E, E2` rather than `S, E2, Response`). Also, we wouldn't need to stick as many bounds everywhere for it.

Reply:
When I try doing this I end up needing 4 parameters `Fut, T, E, E2` and it gives me an error about T being unused, which means I need to shove a phantom data into the ResponseState enum somehow I think but I'm not really sure how best to do so.

Reply:
What is `T` in that case? Previously, `T` was the future type, which appears twice in the enum...I don't see where the 4th param is coming from?